File: frontend/src/lib/folder.ts
import { CreateFolder, DeleteFolder, ListFolders, RenameFolder } from '../../wailsjs/go/desktop/App'

export interface Folder {
  id: string
  name: string
  isDefault: boolean
}

export async function createFolder(name: string): Promise<Folder> {
  const result = await CreateFolder(name)
  return { id: result.id, name: result.name, isDefault: result.isDefault }
}

export async function renameFolder(id: string, name: string): Promise<void> {
  await RenameFolder(id, name)
}

export async function deleteFolder(id: string): Promise<void> {
  await DeleteFolder(id)
}

export async function listFolders(): Promise<Folder[]> {
  const results = await ListFolders()
  return results.map((result) => ({
    id: result.id,
    name: result.name,
    isDefault: result.isDefault,
  }))
}

File: internal/interfaces/desktop/folder.go
package desktop

// FolderResult is the desktop-facing DTO for a Folder.
type FolderResult struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

// CreateFolder creates a new folder named name.
func (a *App) CreateFolder(name string) (FolderResult, error) {
	f, err := a.folder.CreateFolder(a.ctx, name)
	if err != nil {
		return FolderResult{}, err
	}
	return FolderResult{ID: f.ID, Name: f.Name}, nil
}

// RenameFolder renames the folder with the given id.
func (a *App) RenameFolder(id, name string) error {
	return a.folder.RenameFolder(a.ctx, id, name)
}

// DeleteFolder deletes the folder with the given id and every session
// inside it — nothing in the folder survives.
func (a *App) DeleteFolder(id string) error {
	return a.folder.DeleteFolder(a.ctx, id)
}

// ListFolders returns every folder.
func (a *App) ListFolders() ([]FolderResult, error) {
	folders, err := a.folder.ListFolders(a.ctx)
	if err != nil {
		return nil, err
	}
	results := make([]FolderResult, len(folders))
	for i, f := range folders {
		results[i] = FolderResult{ID: f.ID, Name: f.Name}
	}
	return results, nil
}
